CMakeLists.txt

```diff
@@ -298,36 +321,44 @@ endif()
 
 option(USE_OPTIMIZED_LAPACK "Whether or not to use an optimized LAPACK library instead of included netlib LAPACK" OFF)
 
+if(LAPACK)
+
 # --------------------------------------------------
 # LAPACK
-# User did not provide a LAPACK Library but specified to search for one
-if(USE_OPTIMIZED_LAPACK)
-  find_package(LAPACK)
-endif()
 
 # Check the usage of the user provided or automatically found LAPACK libraries
-if(LAPACK_LIBRARIES)
-  include(CheckFortranFunctionExists)
-  set(CMAKE_REQUIRED_LIBRARIES ${LAPACK_LIBRARIES})
-  # Check if new routine of 3.4.0 is in LAPACK_LIBRARIES
-  CHECK_FORTRAN_FUNCTION_EXISTS("dgeqrt" LATESTLAPACK_FOUND)
-  unset(CMAKE_REQUIRED_LIBRARIES)
-  if(LATESTLAPACK_FOUND)
-    message(STATUS "--> LAPACK supplied by user is WORKING, will use ${LAPACK_LIBRARIES}.")
+if(USE_OPTIMIZED_LAPACK)
+  if(LAPACK_LIBRARIES)
+    include(CheckFortranFunctionExists)
+    set(CMAKE_REQUIRED_LIBRARIES ${LAPACK_LIBRARIES})
+    # Check if new routine of 3.4.0 is in LAPACK_LIBRARIES
+    CHECK_FORTRAN_FUNCTION_EXISTS("dgeqrt" LAPACK_FOUND)
+    unset(CMAKE_REQUIRED_LIBRARIES)
+    if(LAPACK_FOUND)
+      message(STATUS "--> LAPACK supplied by user is WORKING, will use ${LAPACK_LIBRARIES}.")
+    else()
+      message(ERROR "--> LAPACK supplied by user is not WORKING or is older than LAPACK 3.4.0, CANNOT USE ${LAPACK_LIBRARIES}.")
+      message(ERROR "-->     Fix LAPACK_LIBRARIES or ensure that USE_OPTIMIZED_LAPACK is off for REFERENCE LAPACK.")
+    endif()
   else()
-    message(ERROR "--> LAPACK supplied by user is not WORKING or is older than LAPACK 3.4.0, CANNOT USE ${LAPACK_LIBRARIES}.")
-    message(ERROR "-->     Will use REFERENCE LAPACK (by default)")
-    message(ERROR "-->     Or Correct your LAPACK_LIBRARIES entry ")
-    message(ERROR "-->     Or Consider checking USE_OPTIMIZED_LAPACK")
+    # Let's hope for any working LAPACK.
+    find_package(LAPACK)
   endif()
 endif()
 
 # Neither user specified or optimized LAPACK libraries can be used
-if(NOT LATESTLAPACK_FOUND)
+if(NOT LAPACK_FOUND)
+  if(USE_OPTIMIZED_LAPACK)
+    message(FATAL_ERROR "USE_OPTIMIZED_LAPACK set but no usable LAPACK provided.")
+  endif()
   message(STATUS "Using supplied NETLIB LAPACK implementation")
   set(LAPACK_LIBRARIES ${LAPACKLIB})
+  set(LAPACK_PCLIBS -l${LAPACKLIB})
   add_subdirectory(SRC)
 else()
+  # This is the full path to a .so file from find_package, possibly.
+  # not ideal. Better ideas?
+  set(LAPACK_PCLIBS ${LAPACK_LIBRARIES})
```

Comment:
Given this expression, the pkg-config file will contain a list of paths to libraries. I do not know if this is portable and I would not expect a list of libraries without `-l` prefix in a pkg-config file. Instead, I would expect pkg-config to return as list of directories and linker flags. That is, instead of
```
-L/tmp/prefix.lapack+openblas/lib -llapack /usr/lib/x86_64-linux-gnu/libopenblas.so
```
I expect
```
-L/usr/lib/x86_64-linux-gnu/openblas-pthread/ -llapack -lopenblas
```

Reply:
I too expect -lfoo instead of a path to an .so file. It is fine in my use-case as I specify build things like this:
```
LAPACK_COMPONENT=       lapacke
LAPACK_COMPONENT_CMAKE_ARGS=    \
        -DUSE_OPTIMIZED_BLAS=ON \
        -DBLAS_LIBRARIES=${BLAS_LIBS:Q} \
        -DUSE_OPTIMIZED_LAPACK=ON \
        -DLAPACK_LIBRARIES=${LAPACK_LIBS:Q} \
        -DCBLAS=OFF -DLAPACKE=ON
# calling cmake with CMAKE_ARGS down the line
```
This means BLAS_LIBS and LAPACK_LIBS containing the linker flags to get the respective libs (usually the same, but possibly simply -lblas and -llapack, or number funky vendor libs). This is what I want in the .pc file.

I do not know how to tell CMake to construct that from libs it found itself is there a function to translate `/path/to/libfoo.so` to `-L/path/to/libfoo -lfoo`? . In my setup, I don't _want_ it to find anything by itself. It should use the libs I tell it to.

The reason for that is also that looking for blas.pc doesn't make sense in a world where a) a .pc (or .cmake) file for the desired BLAS does not exist or b) it is named openblas-xyzprocessor-pthread64.pc or something more weird. For Requires in .pc to make sense, one would have to specify the package name to use during the build. Hardcoding it to 'blas' is just wrong.

Since optimized BLAS libs seem to (optionally) include CBLAS and LAPACKE after all, the combination of these from the netlib build with an optimized BLAS is not on table that often. But I like the idea of having it as building blocks and at least an option of the same level of wrapper API. Eases testing combinations. But this point might become moot once I cave in and build OpenBLAS with CBLAS and LAPACKE directly, while Intel ships it all with MKL in binary anyway.

Maybe the whole idea of making the build more modular to is becoming moot. But I do think that for testing purposes, it is valuable to be able to combine the layers of differing implementations, hence the build and the .pc files being fixed up to refer these.